explorer/src/ingested.rs

```diff
@@ -1,45 +1,134 @@
-use postgres::Client;
+use postgres::{fallible_iterator::FallibleIterator, types::ToSql, Client};
+use shared_lru::*;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
 
-pub fn latest(pg: &mut Client) -> anyhow::Result<Option<u64>> {
-    if let Some(v) = get_key("ingestion/fully_ingested", pg)? {
-        Ok(Some(v.parse()?))
-    } else {
-        Ok(None)
-    }
-}
+use crate::data::{Block, Extrinsic};
 
 fn get_key(key: impl AsRef<str>, pg: &mut Client) -> anyhow::Result<Option<String>> {
-    let row = match pg
-        .query(
+    Ok(pg
+        .query_opt(
             "SELECT value FROM key_values WHERE key = $1",
             &[&key.as_ref()],
         )?
-        .into_iter()
-        .next()
-    {
-        Some(row) => row,
-        None => return Ok(None),
-    };
-
-    Ok(Some(row.get::<_, &str>(&"value").to_string()))
+        .map(|row| row.get::<_, &str>(&"value").to_string()))
 }
 
-pub fn load_block(number: u64, pg: &mut Client) -> anyhow::Result<Option<crate::data::Block>> {
-    if let Some(value) = get_key(format!("block/{}", number), pg)? {
-        Ok(Some(serde_json::from_str(&value)?))
-    } else {
-        Ok(None)
-    }
+pub struct Ingested {
+    blocks: LruCache<u64, Block>,
+    extrinsics: LruCache<(u64, u64), Option<Extrinsic>>,
+    client: Mutex<Client>,
 }
 
-pub fn load_extrinsic(
-    block: u64,
-    index: u64,
-    pg: &mut Client,
-) -> anyhow::Result<Option<crate::data::Extrinsic>> {
-    if let Some(value) = get_key(format!("block/{}/extrinsic/{}", block, index), pg)? {
-        Ok(Some(serde_json::from_str(&value)?))
-    } else {
-        Ok(None)
+impl Ingested {
+    pub fn create(lru: &Arc<SharedLru>, client: Client) -> Ingested {
+        Ingested {
+            blocks: lru.make_cache(),
+            extrinsics: lru.make_cache(),
```

Comment:
Does this mean that both blocks and extrinsics would be cached? I guess that's why you've implemented `MemorySize` to them. My question is where exactly are you using the LRU cache? As I see we are getting specific `u64`s as key, or it would be the least recently used with this key?

Reply:
Yes, both are cached. Calls to `load_block` and `load_extrinsic` pull from the cache first and only do a DB query if the cache is missing the value. The `shared_lru` crate will evict the least recently used item across all caches when full, so any block/extrinsic/something else could be evicted.